File: cpp/Linked List/mergeTwoSortedLists.cpp
/*

Problem Statement: Merge two sorted linked lists and return it as a sorted list. 
The list should be made by splicing together the nodes of the first two lists.

Input: l1 = [1,2,4], l2 = [1,3,4]
Output: [1,1,2,3,4,4]

Input: l1 = [], l2 = [0]
Output: [0]

*/

struct ListNode
{
    int val;
    ListNode *next;
};

ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
{
    ListNode *sortedTemp = new ListNode();
    ListNode *currentHead = sortedTemp;

    while (l1 != NULL && l2 != NULL)
    {
        if (l1->val < l2->val)
        {
            currentHead->next = l1;
            l1 = l1->next;
        }
        else
        {
            currentHead->next = l2;
            l2 = l2->next;
        }

        currentHead = currentHead->next;
    }

    if (l1 != NULL)
    {
        currentHead->next = l1;
        l1 = l1->next;
    }

    if (l2 != NULL)
    {
        currentHead->next = l2;
        l2 = l2->next;
    }

    return sortedTemp->next;
}

File: csharp/blind-75/mergeTwoSortedLists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {
        ListNode sortedHead = new();
        ListNode currentHead = sortedHead;
        
        while (list1 is not null && list2 is not null) {
            if (list1.val < list2.val) {
                currentHead.next = list1;
                list1 = list1.next;
            } else {
                currentHead.next = list2;
                list2 = list2.next;
            }
            
            currentHead = currentHead.next;
        }
        
        if (list1 is not null) {
            currentHead.next = list1;
            list1 = list1.next;
        }
        
        if (list2 is not null) {
            currentHead.next = list2;
            list2 = list2.next;
        }
        
        return sortedHead.next;
    }
}

// Example 1:

// Input: list1 = [1,2,4], list2 = [1,3,4]
// Output: [1,1,2,3,4,4]

// Example 2:

// Input: list1 = [], list2 = []
// Output: []

// Example 3:

// Input: list1 = [], list2 = [0]
// Output: [0]